File: atlas-ai-py/atlas_ai/decision/decision_engine.py
from atlas_ai.types import DecisionContext, Decision


class DecisionEngine:
    def __init__(self, runtime=None):
        self.runtime = runtime

    def decide(self, ctx: DecisionContext) -> list[Decision]:
        event = ctx.event
        decisions: list[Decision] = []

        if event.type == "TICK":
            dt = event.payload.get("dt", 0) if isinstance(event.payload, dict) else 0

            if dt > 40:
                def make_execute(runtime, event):
                    def execute():
                        tid = f"latency-{event.timestamp}"
                        if runtime is not None and hasattr(runtime, "register_task"):
                            runtime.register_task(tid)
                        if runtime is not None and hasattr(runtime, "run_task"):
                            runtime.run_task(tid)
                    return execute

                decisions.append(Decision(
                    name="HighLatencyResponse",
                    confidence=0.8,
                    execute=make_execute(self.runtime, event),
                ))

        return decisions

File: atlas-ai/Decision/DecisionEngine.ts
import { DecisionContext, Decision } from "./types";
import { AtlasRuntime } from "../../atlas-runtime/Lifecycle/AtlasRuntime";
import { BasicTask } from "../../atlas-examples/BasicTask";

export class DecisionEngine {
  private runtime: AtlasRuntime;

  constructor(runtime: AtlasRuntime) {
    this.runtime = runtime;
  }

  decide(ctx: DecisionContext): Decision[] {
    const event = ctx.event;

    const decisions: Decision[] = [];

    // 🔥 RULE 1: system heartbeat reaction
    if (event.type === "TICK") {
      const dt = event.payload.dt;

      if (dt > 40) {
        decisions.push({
          name: "HighLatencyResponse",
          confidence: 0.8,
          execute: () => {
            const task = new BasicTask(
              `latency-${Date.now()}`,
              "Latency Recovery Task"
            );

            this.runtime.registerTask(task);
            this.runtime.runTask(task.id);
          },
        });
      }
    }

    return decisions;
  }
}